### PR TITLE
Added `--dry-run` option for ACL migrate

### LINKED ISSUE #2770: [FEATURE]: Add traceability and optionality to ACL migration.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

The table ACL migration process is opaque and sometimes fails for customers.
We want to allow customers to review the crawl ACLs and make a determination whether or not to upgrade these ACLs.

### Proposed Solution

Log the ACLs that were discovered in the table ACL migration process.
Allow running ACL migration in a "dry run" mode to output the ACLs and not migrate them.
Add option to migrate table workflow to skip ACL migration.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by FastLee**
related to #2770

### CODE REVIEW

File: src/databricks/labs/ucx/cli.py [hunk 1]
```diff
@@ -705,8 +705,14 @@ def migrate_acls(
         workspace_contexts = [ctx]
     else:
         workspace_contexts = _get_workspace_contexts(w, a, run_as_collection, **named_parameters)
-    target_catalog, hms_fed = named_parameters.get("target_catalog"), named_parameters.get("hms_fed", False)
+    target_catalog, hms_fed, dry_run = (
+        named_parameters.get("target_catalog"),
+        named_parameters.get("hms_fed", False),
+        named_parameters.get("dry_run", False),
+    )
```
Comment (nfx) [inline src/databricks/labs/ucx/cli.py:712]:
```suggestion
    target_catalog = named_parameters.get("target_catalog")
    hms_fed = named_parameters.get("hms_fed", False)
    dry_run = named_parameters.get("dry_run", False)
```
nit. every line is a liability, make them less and simpler

File: src/databricks/labs/ucx/cli.py [hunk 2]
```diff
@@ -705,8 +705,14 @@ def migrate_acls(
...
     for workspace_context in workspace_contexts:
+        if dry_run:
+            workspace_context.acl_migrator.snapshot()
```
Comment (nfx) [inline src/databricks/labs/ucx/cli.py:715]: and what do we do with it?

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 3]
```diff
@@ -846,6 +863,20 @@ def migrate_acls(self, *, target_catalog: str | None = None, hms_fed: bool = Fal
             tables_to_migrate = self._get_migrated_tables(tables)
         self._migrate_acls(tables_to_migrate)
 
+    def retrieve_table_acls(self, *, target_catalog: str | None = None, hms_fed: bool = False) -> Iterable[Grant]:
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/grants.py:866]:
```suggestion
    def _retrieve_table_acls(self, *, target_catalog: str | None = None, hms_fed: bool = False) -> Iterable[Grant]:
```
make private

File: src/databricks/labs/ucx/install.py [hunk 4]
```diff
@@ -127,6 +127,7 @@ def deploy_schema(sql_backend: SqlBackend, inventory_schema: str):
             functools.partial(table, "directfs_in_queries", DirectFsAccess),
             functools.partial(table, "used_tables_in_paths", UsedTable),
             functools.partial(table, "used_tables_in_queries", UsedTable),
+            functools.partial(table, "acls", Grant),
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:130]: be more descriptive what it is

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 5]
```diff
@@ -892,9 +923,26 @@ def _get_hms_fed_tables(self, tables: list[Table], target_catalog) -> list[Table
...
         for table in tables_in_scope:
-            tasks.append(partial(self._migrate_grants.apply, table.src, table.rule.as_uc_table_key))
+            tasks.append(partial(self._migrate_grants.apply, table.src, table.rule.as_uc_table))
         Threads.strict("migrate grants", tasks)
 
+    def _retrieve_acls(self, tables_in_scope: list[TableToMigrate]) -> Iterable[Grant]:
+        tasks = []
+        for table in tables_in_scope:
+            tasks.append(partial(self._migrate_grants.retrieve, table.src, table.rule.as_uc_table))
+        grants, errors = Threads.gather("retrieve grants", tasks)
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/grants.py:933]: we don't make any network calls in `retrieve` (after caching), so adding multithreading here increases complexity without any added benefit. rewrite this as synchronous implmentation

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/cli.py (hunk 2 above)
Comment (FastLee) [reply src/databricks/labs/ucx/cli.py:715]: That will be displayed in a dashboard.

File: src/databricks/labs/ucx/hive_metastore/grants.py (hunk 5 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/grants.py:933]: KK

File: src/databricks/labs/ucx/cli.py (hunk 1 above)
Comment (FastLee) [reply src/databricks/labs/ucx/cli.py:712]: Done

File: src/databricks/labs/ucx/install.py (hunk 4 above)
Comment (FastLee) [reply src/databricks/labs/ucx/install.py:130]: Fixed

File: src/databricks/labs/ucx/cli.py (hunk 2 above)
Comment (FastLee) [reply src/databricks/labs/ucx/cli.py:713]: Coming in a follow up PR

Comment (nfx) [reply src/databricks/labs/ucx/cli.py:713]: in the current shape, it just snapshots the data and continues to migrate ACLs - **not doing what this PR intends to do**, introducing a bug.

also user is not presented with any results of a dry run. no log, no whatever.

image

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/cli.py [hunk 6]
```diff
@@ -705,8 +705,16 @@ def migrate_acls(
...
+    target_catalog = named_parameters.get("target_catalog")
+    hms_fed = named_parameters.get("hms_fed", False)
+    dry_run = named_parameters.get("dry_run", False)
     for workspace_context in workspace_contexts:
+        if dry_run:
+            grants = workspace_context.acl_migrator.snapshot()
+            logger.info(
+                f"Dry run completed. Found {len(grants)} grants. The crawled grants can be found in the 'source_table_grants' table. "
+                "No changes were made."
+            )
```
Comment (nfx) [inline src/databricks/labs/ucx/cli.py:717]: this still results in a call to `workspace_context.acl_migrator.migrate_acls`. and technically, it works only for the first workspace out of many. proper dry run has to be in a separate loop.

File: src/databricks/labs/ucx/install.py [hunk 7]
```diff
@@ -127,6 +127,7 @@ def deploy_schema(sql_backend: SqlBackend, inventory_schema: str):
             functools.partial(table, "directfs_in_queries", DirectFsAccess),
             functools.partial(table, "used_tables_in_paths", UsedTable),
             functools.partial(table, "used_tables_in_queries", UsedTable),
+            functools.partial(table, "source_table_grants", Grant),
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:130]: this name is suboptimal. how can we name it better?

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/cli.py (hunk 6 above)
Comment (FastLee) [reply src/databricks/labs/ucx/cli.py:717]: Can you explain?

Comment (FastLee) [reply src/databricks/labs/ucx/cli.py:717]: Got it

File: src/databricks/labs/ucx/install.py (hunk 7 above)
Comment (FastLee) [reply src/databricks/labs/ucx/install.py:130]: OK

File: src/databricks/labs/ucx/cli.py [hunk 8]
```diff
@@ -705,7 +705,17 @@ def migrate_acls(
...
         workspace_contexts = _get_workspace_contexts(w, a, run_as_collection, **named_parameters)
-    target_catalog, hms_fed = named_parameters.get("target_catalog"), named_parameters.get("hms_fed", False)
+    target_catalog = named_parameters.get("target_catalog")
+    hms_fed = named_parameters.get("hms_fed", False)
+    dry_run = named_parameters.get("dry_run", False)
+    if dry_run:
+        for workspace_context in workspace_contexts:
+            grants = workspace_context.acl_migrator.snapshot()
+        logger.info(
+            f"Dry run completed. Found {len(grants)} grants. The crawled grants can be found in the 'hms_table_access' table. "
```
Comment (nfx) [inline src/databricks/labs/ucx/cli.py:715]: this is incorrect again - you're overwriting `grants` for each workspace.

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 9]
```diff
@@ -828,18 +842,21 @@ def _replace_account_group(self, grant: Grant) -> Grant:
...
         migration_status_refresher: TableMigrationStatusRefresher,
         migrate_grants: MigrateGrants,
+        backend: SqlBackend,
+        schema: str,
     ):
         self._table_crawler = tables_crawler
         self._workspace_info = workspace_info
         self._migration_status_refresher = migration_status_refresher
         self._migrate_grants = migrate_grants
+        super().__init__(backend, "hive_metastore", schema, "hms_table_access", Grant)
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/grants.py:859]:
```suggestion
        super().__init__(backend, "hive_metastore", schema, "inferred_grants", Grant)
```

File: src/databricks/labs/ucx/cli.py (hunk 8 above)
Comment (nfx) [reply src/databricks/labs/ucx/cli.py:715]: also - add a link to `{self._ws.config.host}/explore/data/hive_metastore/{inventory_database}/inferred_grants`

Comment (FastLee) [reply src/databricks/labs/ucx/cli.py:715]: Fixed

File: src/databricks/labs/ucx/cli.py [hunk 10]
```diff
@@ -705,7 +705,22 @@ def migrate_acls(
...
+        total_grants = 0
+        for workspace_context in workspace_contexts:
+            grants = workspace_context.acl_migrator.snapshot()
+            total_grants += len(grants)
+        logger.info(
+            f"Dry run completed. Found {total_grants} grants. The crawled grants can be found in the 'inferred_grants' table. "
+            f"The URL for the grants table: {w.config.host}/explore/data/hive_metastore/"
+            f"{current_ws_context.config.inventory_database}/inferred_grants "
+            "No changes were made."
+        )
```
Comment (nfx) [inline src/databricks/labs/ucx/cli.py:722]:
```suggestion
            grants = workspace_context.acl_migrator.snapshot()
            total_grants += len(grants)
            logger.info(
            f"Found {total_grants} grants. The crawled grants can be found in the 'inferred_grants' table. "
            f"The URL for the grants table: {workspace_context.workspace_client.config.host}/explore/data/hive_metastore/"
            f"{current_ws_context.config.inventory_database}/inferred_grants "
            "No changes were made."
        )
```
do it per workspace

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/cli.py (hunk 10 above)
Comment (FastLee) [reply src/databricks/labs/ucx/cli.py:722]: OK

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**